homeassistant/components/utility_meter/sensor.py

```diff
@@ -355,53 +377,70 @@ def start(self, unit):
         self._state = 0
         self.async_write_ha_state()
 
+    @staticmethod
+    def _validate_state(state: State | None) -> Decimal | None:
+        """Parse the state as a Decimal if available. Throws DecimalException if the state is not a number."""
+        return (
+            None
+            if state is None or state.state in [STATE_UNAVAILABLE, STATE_UNKNOWN]
+            else Decimal(state.state)
+        )
+
+    def calculate_adjustment(
+        self, old_state: State | None, new_state_val: Decimal
+    ) -> Decimal | None:
+        """Calculate the adjustment based on the old and (assumed valid) new state."""
```

Comment:
based on the description of the method, maybe they should both be `State` as in `new_state.value` ?(just trying to make things clearer)

Reply:
Yes and no, in order to get both values, you have to parse both states as a decimal while you might not need both values (in the case of delta values or the new periodically resetting option), which might raise DecimalExceptions if the old state is invalid.

I agree that both argumetns should be of the same type to make it clearer, I will think about another solution.

Reply:
Yes true, but if the delta_values are used, we are only interested in the new_state and not in the old state. If we, before calling this calculate_adjustment method already check if both the old and new state are valid (because that is needed in order to get the Decimal value of a state), we might add an error log that the old state is invalid while we are not even interested in the old state.

Another option is to do Decimal parsing for both the old_state and new_state in the `calculate_adjustment` method but than we cannot already return once we discover that the new_state is invalid and continuing is useless, as I do now.

Reply:
check if it's delta or not and parse to Decimal accordingly in `calculate_adjustment` ?

Reply:
I've changed both parameters to the `State` object. The old_state will, if necessary be parsed inside `calculate_adjustment` wrapped in a try/ except to catch parsing errors. The new_state is parsed inside `calculate_adjustment` as well including a try/ except. While this new_state will never fail to parse (because it is already validated before the `calculate_adjustment` is called), it is safer for the future, if the `calculate_adjustment` may be re-used somewhere else.